### PR TITLE
feat: create trpc route to fetch download link

### LINKED ISSUE #7488: [CAL-1170] Some users are not able to download recording
Found a bug? Please fill out the sections below. 👍

### Issue Summary

Screenshot 2023-03-03 at 2 32 23 AM

<sub>[CAL-1170](https://linear.app/calcom/issue/CAL-1170/some-users-are-not-able-to-download-recording)</sub>

### PR DESCRIPTION
**Pull request by Udit-takkar**
## What does this PR do?

Fixes https://github.com/calcom/cal.com/issues/7488

this was the room name whose recording was not downloadable.

https://user-images.githubusercontent.com/53316345/222750080-786c119f-b834-430d-8d54-ec358433f7dd.mov

**Environment**: Staging(main branch) / Production

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: packages/trpc/server/routers/viewer.tsx [hunk 1]
```diff
@@ -1167,6 +1172,33 @@ const loggedInViewerRouter = router({
...
+        });
+      }
+
+      try {
+        const res = await getDownloadLinkOfCalVideoByRecordingId(recordingId);
+        return res;
+      } catch (err) {
+        throw new TRPCError({
+          code: "BAD_REQUEST",
+        });
```
Comment (sean-brydon) [inline packages/trpc/server/routers/viewer.tsx:1199]: Should we provide more detail if we ever hit this? err.message maybe? (Not sure of the error thrown here could be differnt)

File: packages/core/videoClient.ts [hunk 2]
```diff
@@ -175,6 +175,7 @@ const getRecordingsOfCalVideoByRoomName = async (
   try {
     dailyAppKeys = await getDailyAppKeys();
   } catch (e) {
+    console.error("Error: Cal video provider is not installed.");
```
Comment (emrysal) [inline packages/core/videoClient.ts:178]: Added an error so it's clear when this happens.

File: packages/features/ee/video/ViewRecordingsDialog.tsx [hunk 3]
```diff
@@ -40,116 +41,155 @@ interface GetTimeSpanProps {
...
     hour: "numeric",
     minute: "numeric",
-    hour12: isTimeFormatAMPM,
+    hour12,
   }).format(new Date(endTime));
 
   return `${formattedStartTime} - ${formattedEndTime}`;
 };
 
+const useRecordingDownload = () => {
```
Comment (emrysal) [inline packages/features/ee/video/ViewRecordingsDialog.tsx:65]: Hook specifically for recording downloads.

File: packages/features/ee/video/ViewRecordingsDialog.tsx [hunk 4]
```diff
@@ -40,116 +41,155 @@ interface GetTimeSpanProps {
...
+          window.location.href = data.download_link;
+        }
+      },
+    }
+  );
+
+  return {
+    setRecordingId: (newRecordingId: string) => {
+      // may be a way to do this by default, but this is easy enough.
+      if (recordingId === newRecordingId && data) {
```
Comment (emrysal) [inline packages/features/ee/video/ViewRecordingsDialog.tsx:88]: Possibly can be improved with react query caching and refetch, but couldn't figure it out.

File: packages/features/ee/video/ViewRecordingsDialog.tsx [hunk 5]
```diff
@@ -40,116 +41,155 @@ interface GetTimeSpanProps {
...
-  };
-
   const subtitle = `${booking?.title} - ${dayjs(booking?.startTime).format("ddd")} ${dayjs(
     booking?.startTime
   ).format("D")}, ${dayjs(booking?.startTime).format("MMM")} ${getTimeSpan({
     startTime: booking?.startTime,
     endTime: booking?.endTime,
     locale: i18n.language,
-    isTimeFormatAMPM: timeFormat === 12,
+    hour12: timeFormat === 12,
```
Comment (emrysal) [inline packages/features/ee/video/ViewRecordingsDialog.tsx:173]: Consistency with elsewhere

File: packages/features/ee/video/ViewRecordingsDialog.tsx [hunk 6]
```diff
@@ -40,116 +41,155 @@ interface GetTimeSpanProps {
...
+      ) : (
+        (!recordings || (recordings && "total_count" in recordings && recordings?.total_count === 0)) && (
+          <p className="font-semibold">{t("no_recordings_found")}</p>
+        )
+      )}
+    </>
+  );
+};
+
 export const ViewRecordingsDialog = (props: IViewRecordingsDialog) => {
```
Comment (emrysal) [inline packages/features/ee/video/ViewRecordingsDialog.tsx:157]: Decreasing the amount of rerenders of this component improves overal performance - setIsOpenDialog rerenders the parent too and all children.

**Review by emrysal — APPROVED** [review]
Tested everything through from my end, made some improvements here and there & works!

File: packages/features/ee/video/ViewRecordingsDialog.tsx (hunk 4 above)
Comment (Udit-takkar) [reply packages/features/ee/video/ViewRecordingsDialog.tsx:88]: @emrysal this looks a bit confusing.  
I have set the cacheTime to 0 so even if you click same download recording button onSuccess would get trigger and  set window.location.href = data.download_link

Comment (Udit-takkar) [reply packages/features/ee/video/ViewRecordingsDialog.tsx:88]: we just have to set recording id and that should work

**Pull request merged**